### PR TITLE
Gherkin-Utils - Java - Address issue #1662 - add a Java pretty print gherkin module

### PR DESCRIPTION
**Pull request by nddipiazza**
## Summary

Ported the TypeScript Gherkin pretty print utility to Java.

## Details

Make a java port of the javascript pretty printing library available in gherkin-utils.

## Motivation and Context

A huge amount of the Gherkin users out there are java users through cucumber-jvm project. We need a pretty printer for Gherkin.

## How Has This Been Tested?

Took most of the unit tests from the typescript project.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue).
- [x] New feature (non-breaking change which adds functionality).
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected).

## Checklist:

- [x] The change has been ported to Java.
- [x] I've added tests for my code.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have updated the CHANGELOG accordingly.

### CODE REVIEW

**Comment by aslakhellesoy** [comment]
Wow thanks so much for contributing this! I haven't had a chance to look at it yet, but will do my best to do it this week.

File: gherkin-utils/java/.rsync [hunk 1]
```diff
@@ -0,0 +1,6 @@
+../LICENSE LICENSE
+../../.templates/github/ .github/
+../../.templates/java/ .
+../testdata/ testdata/
+../gherkin.berp gherkin.berp
```
Comment (aslakhellesoy) [inline gherkin-utils/java/.rsync:5]: Please remove this line - it's for the gherkin library only

File: gherkin-utils/java/.subrepo [hunk 2]
```diff
@@ -0,0 +1 @@
+cucumber/gherkin-java
```
Comment (aslakhellesoy) [inline gherkin-utils/java/.subrepo:1]: Please remove this file

File: gherkin-utils/java/.rsync [hunk 3]
```diff
@@ -0,0 +1,6 @@
...
+../gherkin-languages.json src/main/resources/io/cucumber/gherkin/gherkin-languages.json
```
Comment (aslakhellesoy) [inline gherkin-utils/java/.rsync:6]: Remove this line too.

File: gherkin-utils/java/LICENSE [hunk 4]
```diff
@@ -0,0 +1,21 @@
+The MIT License (MIT)
+
+Copyright (c) Cucumber Ltd, Gaspar Nagy, Björn Rasmusson, Peter Sergeant
```
Comment (aslakhellesoy) [inline gherkin-utils/java/LICENSE:3]: Only `Cucumber Ltd` please, and optionally your own name.

File: gherkin-utils/java/Makefile [hunk 5]
```diff
@@ -0,0 +1,53 @@
+include default.mk
+
```
Comment (aslakhellesoy) [inline gherkin-utils/java/Makefile:2]: Please remove everything below this line.

File: gherkin-utils/java/README.md [hunk 6]
```diff
@@ -0,0 +1,7 @@
+[![Maven Central](https://img.shields.io/maven-central/v/io.cucumber/gherkin.svg?label=Maven%20Central)](https://search.maven.org/search?q=g:%22io.cucumber%22%20AND%20a:%22gherkin%22)
```
Comment (aslakhellesoy) [inline gherkin-utils/java/README.md:1]: Please fix the full contents of this file (replace gherkin with gherkin-utils). Fix the content.

File: gherkin-utils/java/gherkin-java.razor [hunk 7]
```diff
@@ -0,0 +1,320 @@
+//      This code was generated by Berp (http://https://github.com/gasparnagy/berp/).
```
Comment (aslakhellesoy) [inline gherkin-utils/java/gherkin-java.razor:1]: Please remove this file

File: gherkin-utils/java/gherkin.berp [hunk 8]
```diff
@@ -0,0 +1,38 @@
+[
```
Comment (aslakhellesoy) [inline gherkin-utils/java/gherkin.berp:1]: Please remove this file

File: gherkin-utils/java/pom.xml [hunk 9]
```diff
@@ -0,0 +1,41 @@
+<?xml version="1.0"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+  <modelVersion>4.0.0</modelVersion>
+  <groupId>io.cucumber</groupId>
+  <artifactId>gherkin-utils</artifactId>
+  <version>21.0.1-SNAPSHOT</version>
```
Comment (aslakhellesoy) [inline gherkin-utils/java/pom.xml:7]: Please align the version number with the existing `gherkin-utils/javascript` version (and add 1 for the patch version plus a `-SNAPSHOT` suffix).

File: gherkin-utils/java/pom.xml [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+  <scm>
+    <connection>scm:git:git://github.com/cucumber/common.git</connection>
+    <developerConnection>scm:git:git@github.com:cucumber/common.git</developerConnection>
+    <url>git://github.com/cucumber/common.git</url>
+    <tag>HEAD</tag>
+  </scm>
+
+  <dependencies>
+    <dependency>
+      <groupId>io.cucumber</groupId>
```
Comment (aslakhellesoy) [inline gherkin-utils/java/pom.xml:29]: This should be `test` scope.

Please add a runtime dependency on `messages`

File: gherkin-utils/java/src/main/java/io/cucumber/gherkin/util/PrettyPrintGherkinDocument.java [hunk 11]
```diff
@@ -0,0 +1,371 @@
...
+                keywordPrefix(level, syntax) +
+                stepContainer.getKeyword() +
+                ": " +
+                stepContainer.getName() +
+                "\n" +
+                description +
+                (description != null && !"".equals(description.trim()) && stepCount > 0 ? "\n" : "");
+    }
+
+    private static String prettyKeywordContainer(
```
Comment (aslakhellesoy) [inline gherkin-utils/java/src/main/java/io/cucumber/gherkin/util/PrettyPrintGherkinDocument.java:149]: There is a lot of duplication in this method and the next three. Can you extract a private method that they all use?

**Review by aslakhellesoy — CHANGES_REQUESTED** [review]
Overall great code! I've left some comments, mostly about deleting unnecessary files that aren't needed.

Also please add an entry to the CI build configuration in `.circleci/config.yml` to make sure this gets built.

**Comment by jamietanna** [comment]
As mentioned in https://github.com/cucumber/common/issues/1729 I've literally just finished working on this as an external project 😭 (sorry for not raising an issue to share that I was!)

So I'd be able to help it you'd like with implementation / we can see about pulling my library in here?

File: gherkin-utils/java/src/test/java/io/cucumber/gherkin/util/PrettyPrintGherkinDocumentTest.java [hunk 12]
```diff
@@ -0,0 +1,283 @@
...
+    }
+
+    @Test
+    public void emptyFile() {
+        GherkinDocument gherkinDocument = parser.parse("");
+        assertEquals("", PrettyPrintGherkinDocument.prettyPrint(gherkinDocument, Syntax.gherkin));
+    }
+
+    @Test
+    public void scenarioOutlineWithExamples() {
```
Comment (jamietanna) [inline gherkin-utils/java/src/test/java/io/cucumber/gherkin/util/PrettyPrintGherkinDocumentTest.java:29]: To make these test classes more readable, we could split the cases into files before/after, then assert on that, which means we don't have to escape / deal with long strings in this test class

File: gherkin-utils/java/pom.xml [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+      <artifactId>gherkin</artifactId>
+      <version>21.0.0</version>
+    </dependency>
+
+    <dependency>
+      <groupId>junit</groupId>
+      <artifactId>junit</artifactId>
```
Comment (jamietanna) [inline gherkin-utils/java/pom.xml:36]: :question: are we still as a project using JUnit4? Or are we able to update this to JUnit5?

Comment (aslakhellesoy) [reply gherkin-utils/java/pom.xml:36]: We can move to junit5, we use that in a few other places

**Comment by aslakhellesoy** [comment]
> As mentioned in #1729 I've literally just finished working on this as an external project 😭 (sorry for not raising an issue to share that I was!)
> 
> So I'd be able to help it you'd like with implementation / we can see about pulling my library in here?

That would be great @jamietanna! Just keep in mind that we aim to keep the TypeScript/Java implementations consistent.

**Review by jamietanna — COMMENTED** [review]
Can we please make sure this is testing against everything in `gherkin/testadata` - looks like `escaped_pipes.feature` returns a slightly different result that I'd expect (`\\o` vs `\o`) and there are likely a few other edge cases

**Comment by aslakhellesoy** [comment]
> Can we please make sure this is testing against everything in gherkin/testadata - looks like escaped_pipes.feature returns a slightly different result that I'd expect (\\o vs \o) and there are likely a few other edge cases

The files in `gherkin/testadata` are (intentionally) not formatted according to the `pretty` implementation. We can read them all in and format them (the tests in the TypeScript implementation does this), but it doesn't compare the formatted document against anything - we just verify that we can parse the formatted document again.

I think the best way to ensure edge cases are covered is to write dedicated unit tests where we have control over the expected output.

**Comment by nddipiazza** [comment]
OK do you have a list of the test feature files somewhere you want me to add to the unit tests? I'm happy to do so. 
I probably won't finish these updates requested for a few days, doing it in my spare time as my day job is eating up my day on other fires.

**Comment by jamietanna** [comment]
Not to worry @nddipiazza!

You can use https://gitlab.com/jamietanna/gherkin-formatter/-/tree/main/src/test/resources/input as the inputs, and https://gitlab.com/jamietanna/gherkin-formatter/-/tree/main/src/test/resources/output as outputs (for passing tests)

Which are following a formatting style I've found useful on my own implementation.

These were largely adapted from https://github.com/cucumber/gherkin-java/tree/master/testdata but I added quite a few additional scenarios as it was required to make sure I had everything really covered.

**Comment by nddipiazza** [comment]
Excellent. I'm on it.

**Comment by jamietanna** [comment]
Feel free to copy code / thinking from the library, too - it's MIT so we can bring it into this project as needed.

There's some additional complexities that don't appear to be in the Typescript version (like comments) so it may be there's a bit of extra work rather than just a plain port :sweat_smile:

**Comment by aslakhellesoy** [comment]
Yes, comments are lost in the current implementation, and it would be great to keep them! I think we can do that by popping them off as we traverse the other nodes, comparing line numbers.

**Comment by jamietanna** [comment]
Comments seem to be the most difficult piece in the puzzle right now - I'm still trying to get my implementation to handle them right 😥

**Comment by aslakhellesoy** [comment]
I've submitted draft PR #1738 to pretty print comments. Feedback welcome!

**Comment by laeubi** [comment]
@aslakhellesoy @jamietanna can you tell when/where this will finally be released? This could become very handy for the cucumber-eclipse plugin as it currently lacks formatting capabilities.

**Comment by jamietanna** [comment]
@laeubi in https://www.jvt.me/posts/2021/09/07/gherkin-java-formatter/ I've released a formatter for Gherkin but unfortunately comments are very tricky and I've not quite solved it there yet.

You're welcome to see how it works with your scenarios, but it's not ideal right now 😥

**Comment by laeubi** [comment]
> You're welcome to see how it works with your scenarios, but it's not ideal right now disappointed_relieved

Is this released as a maven artifact somewhere or should i simply copy the code into the eclipse-cucumber project for now?

**Comment by jamietanna** [comment]
Nice I'll try and have a look later - check https://gitlab.com/jamietanna/gherkin-formatter/-/merge_requests/3 out for some additional edge cases I've run into ☺

File: gherkin-utils/java/pom.xml (hunk 10 above)
Comment (nddipiazza) [reply gherkin-utils/java/pom.xml:29]: i could not make this change, it caused project not to compile due to missing dependencies during compile time.

**Comment by aurelien-reeves** [comment]
FYI, in case you missed it, here's the error in the CI:

```console
touch .rsynced
[[ -z $(git status -s) ]] || ( \
	echo "Working copy is dirty. Please run 'source scripts/functions.sh && rsync_files' and commit modified files." && \
	echo "Found: " && \
	git status -s && \
	exit 1 \
)
Working copy is dirty. Please run 'source scripts/functions.sh && rsync_files' and commit modified files.
Found: 
 M gherkin-utils/java/maven-versions-rules.xml
?? gherkin-utils/java/.github/
?? gherkin-utils/java/scripts/
make: *** [Makefile:41: check_synced] Error 
```

**Comment by nddipiazza** [comment]
Oh gosh, I'm finally getting back to this. I got mixed up with other work I was doing and thought this was merged. 

I fixed the comment that was pending and will look into the failing test.

**Comment by aslakhellesoy** [comment]
Thanks for picking this up! I’ll let you know when I’ve had a chance to review it.

**Comment by nddipiazza** [comment]
@jamietanna @aslakhellesoy @aurelien-reeves ready for review. 

want me to add any more test cases?

File: .templates/java/scripts/check-jar.sh [hunk 14]
```diff
@@ -8,11 +8,11 @@ set -uf -o pipefail
...
   jar="$1"
   module_name=$(xmlstarlet sel -N pom="http://maven.apache.org/POM/4.0.0"  -t -m "//pom:project.Automatic-Module-Name" -v . pom.xml)
-  module_path=$(echo $module_name | sed "s/\./\\\\\//g")
-  echo "Checking contents of ${jar}"
+  module_path=$(echo $module_name | sed "s/\./\\\\\//g" | sed "s/-/\\\\\//g")
+  echo "Checking contents of ${jar} to see if it matches pattern: ${module_path}"
   unshaded_classes=$(unzip -l ${jar} | grep -e "\.class" | rev | cut -d' ' -f1 | rev | grep -v "^$module_path")
   if [[ "${unshaded_classes}" != "" ]]; then
-    echo "Some classes in ${jar} are not in the io.cucumber package. Rename the classes or change the maven-shade-plugin configuration."
+    echo "Some classes in ${jar} are not in the expected package matching pattern ${module_path}. Rename the classes or change the maven-shade-plugin configuration."
```
Comment (aurelien-reeves) [inline .templates/java/scripts/check-jar.sh:15]: Could you explain those changes here?

**Review by aurelien-reeves — CHANGES_REQUESTED** [review]
From my point of view, it looks good.

One entry in the changelog of `gherkin-utills` is missing to document the addition of that new java implementation :p

I would just have one question regarding the changes in `check-jar.sh` (see within the file)

And also, it would be a great addition to update all the readmes of the component to:
- have a global introduction to gherkin-utils in the root of that component
- have dedicated instructions for the javascrit and the java implementations

However, I would agree to report that as another piece of work in order to move-on with the current PR. Please, just at least open an issue, or even better, a draft PR, to keep track of that.

For the java code, I'll let other to take a look 😅

File: .templates/java/scripts/check-jar.sh (hunk 14 above)
Comment (nddipiazza) [reply .templates/java/scripts/check-jar.sh:15]: As I added the new module, I noticed that the message was misleading. It mentioned "io.cucumber" needs to be the package, but it's actually checking for io.cucumber + {sub-module}. After spinning my tires on that, I updated the message to make it clear.

And sub-modules with `-` in them need to be replaced with `/` just like you are already doing with `.`

**Comment by aslakhellesoy** [comment]
I've continued this work in #1985 - closing this.
